src/FastRouteRouter.php

```diff
@@ -277,41 +255,95 @@ public function generateUri($name, array $substitutions = [], array $options = [
         }
 
         $route   = $this->routes[$name];
-        $path    = $route->getPath();
         $options = ArrayUtils::merge($route->getOptions(), $options);
 
         if (! empty($options['defaults'])) {
             $substitutions = array_merge($options['defaults'], $substitutions);
         }
 
-        foreach ($substitutions as $key => $value) {
-            $pattern = sprintf(
-                '~%s~x',
-                sprintf(self::VARIABLE_REGEX, preg_quote($key))
-            );
-            $path = preg_replace($pattern, $value, $path);
-        }
+        $routeParser       = new RouteParser();
+        $routes            = array_reverse($routeParser->parse($route->getPath()));
+        $missingParameters = [];
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($routes as $parts) {
+            // Check if all parameters can be substituted
+            $missingParameters = $this->missingParameters($parts, $substitutions);
 
-        // 1. remove optional segments' ending delimiters
-        //    and remove leftover regex char classes like `:[` and `:prefix-[` (issue #18)
-        // 2. split path into an array of optional segments and remove those
-        //    containing unsubstituted parameters starting from the last segment
-        $path = preg_replace('/\]|:.*\[/', '', $path);
-        $segs = array_reverse(explode('[', $path));
-        foreach ($segs as $n => $seg) {
-            if (strpos($seg, '{') !== false) {
-                if (isset($segs[$n - 1])) {
-                    throw new Exception\InvalidArgumentException(
-                        'Optional segments with unsubstituted parameters cannot '
-                        . 'contain segments with substituted parameters when using FastRoute'
-                    );
+            // If not all parameters can be substituted, try the next route
+            if (! empty($missingParameters)) {
+                continue;
+            }
+
+            // Generate the path
+            $path = '';
+            foreach ($parts as $part) {
+                if (is_string($part)) {
+                    // Append the string
+                    $path .= $part;
+                    continue;
+                }
+
+                // Check substitute value with regex
+                $regex = '~^' . str_replace('/', '\/', $part[1]) . '$~';
```

Comment:
I'm not 100% clear on why the `str_replace` is necessary here. If we're adding `~` to the start and end, `preg_match` doesn't need the forward slashes escaped. For instance, if `$part[1]` is `foo/bar/baz`, then all of the following return `1`:

* `preg_match('/^foo\/bar\/baz$/', $part[1]);`
* `preg_match('~^foo\/bar\/baz$~', $part[1]);`
* `preg_match('~^foo/bar/baz$~', $part[1]);`

If I haven't misunderstood something, then I suggest we remove line `288` and replace the next line with:

```php
if (! preg_match('~^' . $part[1] . '$~', $substitutions[$part[0]])) {
```